### ISSUE TITLE
SFML claims to be DPI aware on Windows, when it isn't

### ISSUE THREAD

**Issue by vvuk** [SFML/SFML#1299]
SFML calls `SetProcessDpiAwareness(ProcessSystemDpiAware)`, but then does nothing else to actually handle DPI scaling, or even provide a way to access the global/screen DPI to code that uses SFML.  I would suggest it needs to do on or more of the following:

# Add DPI to the mode struct, so that when the system mode is obtained the DPI is included
# Set a default view on windows that does scaling so that games can assume 96dpi and get properly scaled
# Or just allow apps to opt in to DPI awareness (e.g. being able to declare being unaware, and let the OS take care of scaling)

**Comment by eXpl0it3r** [comment]
The current behavior is as far as I know intentional. However I do also agree that it's not the best situation to be in.

See [this forum discussion](https://en.sfml-dev.org/forums/index.php?topic=17092.0) and maybe #679 might be related as well.

**Comment by eXpl0it3r** [comment]
I'm kind of hijacking this issue, so we have a tracking feature item.

The suggested API change would be:

- `sf::Window::getSize()` provides the OS-scaled window size (as it does right now)
- `sf::Window::getFramebufferSize()` provides the unscaled framebuffer size

With this setup, everyone is able to calculate the difference and adjust accordingly.

Open questions:

- [ ] Do we still set the DPI awareness on Windows?
- [ ] What do we have to change w.r.t. DPI awareness on macOS?

**Comment by MarioLiebisch** [comment]
If DPI Awareness is not declared, both functions would probably return the very same thing, so this only works if SFML declares awareness, but then it's theoretically very inconsistent and misleading regarding values returned, e.g. you set your window to 640x480 and suddenly get back 1280x960 due to DPI scaling. So overall I'd still prefer just keeping the classic behavior (not declaring awareness) and instead add an API to retrieve screen DPI. On Mac it's a different thing, because as far as I remember it will just upscale everything, if not aware, including the title bar and stuff (although I haven't checked that).

**Comment by orthoxerox** [comment]
I agree with @eXpl0it3r. SDL2, for example, distinguishes between `SDL_GetWindowSize()` which returns the size of the client area in scaled pixels and `SDL_GL_GetDrawableSize()` or `SDL_GetRendererOutputSize()` which return the size of the client area in 1:1 pixels. The only addition I would want is for `VideoMode` (or the constructor of `Window`) to include DPI awareness as an explicit parameter.

**Comment by lapinozz** [comment]
Looking at the DPI stuff on windows,

I think the main issue is, what's the most graceful way to handle when the user doesn't deal with DPI

"Dealing" with DPI could be as simple as setting your view's size to the window's "scaled" size instead of it's screen-pixel size, but I wouldn't expect most user to do that if they are not aware of what DPI is or have a high-DPI screen and it can cause other rendering issues.

The first way to handle this is to leave the process as DPI-unaware, the window automatically gets scaled up by Windows and looks as it should, the major downside is that you are rendering at a lower resolution and Windows resizes the content of your window which might make stuff look blurry

The second option is to declare the process as DPI-aware and do nothing special, like we have right now, the main issue is that if you have text elements they might be too small to read and your UI in general might look tiny

The third option is to declare the process as DPI-aware but manually scale sizes and position when handing them between the user and the Win32 API, the advantage is that things would appear scaled-up but would be rendering at a the proper resolution (essentially the equivalent scaling the view to match the DPI to match the presentation size of the window rather than its pixel size). The downsize is that you're rendering at a different resolution than the user expects so text might still look a bit blurry and it can cause artifacts in things like tile maps. It's also more effort and changes on the SFML size to have everything scaled properly.

I would tend more toward the first option, setting the process as DPI-unaware by default and having an option for users to opt-in to DPI awareness

When we do enable DPI aware, either as a default or as an opt-in, we should use the PerMonitorV2 mode just like @texus did in their branch https://github.com/SFML/SFML/compare/master...texus:SFML:feature/high-dpi-windows
This ensure that the window's title bar is scaled properly.
Just like in said branch, we can provide a per-window option to auto-scale the window to match the DPI, so that the size is adjusted on creation depending on the monitor's DPI and changes if it's moved to a different monitor. That seems to be the behavior of Windows modern/correct app.
I'm not sure it's necessary to provide this option, it's something the user can do if they want especially if we have a monitor API so they can query the DPI before creating the window.

Also, DPI-awareness is per-process (well actually per thread but we probably don't want to deal with it at that level) not per-window, so if it's an opt-in option it'd make it a static function on sf::Window you call before creating any window rather than something you pass to sf::Window's constructor/create function

**Comment by lapinozz** [comment]
Testing without DPI-awarness on windows 11 in fullscreen I kinda get the same result as #679 but kinda not. When the window is focused it doesn't get scaled so it's displayed properly but when out of focus it gets scaled up so it becomes bigger than the screen and you can't see part of the window.

**Comment by lapinozz** [comment]
WIP for opt-in DPI-awarness
https://github.com/SFML/SFML/compare/master...lapinozz:SFML:high-dpi-opt-in

Basically just a static function you have to call before opening a window if you want high-DPI
```cpp
sf::Window::setDpiAware();
```

also added `sf::Event::DpiChanged`
